src/python/pants/util/objects.py

```diff
@@ -285,6 +289,14 @@ def __new__(cls, value):
     # won't call this __eq__ (and therefore won't raise like we want).
     def __eq__(self, other):
       """Redefine equality to avoid accidentally comparing against a non-enum."""
+      # Options parsing will use == to check whether an option's value has changed when recording
+      # where the option's value was set (in pants.ini, the command line, the environment). If the
+      # option isn't set and falls back to the default, it will be compared against None. Limiting
+      # __eq__ to raise on some comparisons is intended to avoid accidental comparisons against
+      # e.g. a string not wrapped in an enum instance -- comparing to None is much less confusing
+      # and not clearly an error, so we just return False here.
```

Comment:
I think I see what you're trying to say, but got a bit confused between the `None` part and the `raise` part. Consider tweaking the sentence order to this:

```python
# Options parsing will use == to check whether an option's value has changed when recording
# where the option's value was set (in pants.ini, the command line, the environment). It raises on certain comparisons to avoid accidental comparisons, e.g. a string not being wrapped in an enum instance.
# Note if the option isn't set and it falls back to the default, then the comparison will be against `None`. Comparing # to None is not clearly an error, so we just return False here.
```

Reply:
I've modified it to now read:
```python
      # This __eq__ implementation raises on anything that's not an instance of this enum class to
      # avoid accidental comparisons, e.g. a string not being wrapped in an enum instance. However,
      # options parsing will use == to check whether an option's value has changed when recording
      # where the option's value was set (in pants.ini, the command line, the environment). If the
      # option isn't set and it falls back to the default, then there will be a comparison against
      # None. Comparing an enum to None is not clearly an error, so we just return False here.
```

Let me know if that's more clear!

Reply:
Much better! The `However` still confuses me though, as `will use ==` seems like the normal default case and I would expect `however` to be setting up you describing some special case. I'd remove the `however` and say something like `If the types are valid, options parsing will use ==`.

Reply:
(I spent a few minutes debating whether to add the "however" -- I can take it out again!)

Reply:
I keep going back and forth here -- turned it into:
> Currently, options parsing will use ==

Reply:
Do you even need this comment at all? It seems like a lot of verbiage in order to justify something that doesn't need justification. Allowing comparisons to `None` is an extremely reasonable thing to do. 

In fact, it's ~in violation of the contract of `__eq__` to raise an exception at all. I think you're supposed to return the `NotImplemented` singleton.  But that's another issue, I'm fine with raising for extra caution, but that's the only thing that needs explaining, not why you allow None.

My issue with comments like this is that they reference implementation details of faraway code, and will not get updated when that faraway code changes. So better to just be rid of it entirely. No one will think it's odd that you allow comparisons to None.

Reply:
This sounds great -- I will add a note about looking into better conforming to the enum contract as well.

Reply:
Done!